src/settings.c

```diff
@@ -365,6 +367,62 @@ int read_config(char *config_name, settings_t *settings)
     return 0;
 }
 
+int read_database(char *database_name, settings_t *settings)
+{
+    json_error_t error;
+    const char *section;
+    size_t index;
+    json_t *j_value;
+    json_t *j_entry;
+
+    json_t *database_json = json_array();
+
+    database_json = json_load_file(database_name, 0, &error);
+
+    if (database_json == NULL)
+    {
+        fprintf(stderr, "%s:%d:%d error:%s \n",
+                database_name, error.line, error.column, error.text);
+        return 1;
+    }
+
+    json_array_foreach(database_json, index, j_entry)
+    {
+        device_database_t *entry = (device_database_t*)malloc(sizeof(device_database_t));
+        entry->next = settings->coap.security;
```

Comment:
If malloc fails, this causes segfault. Fail fast.
If malloc doesn't fail, entry should be pre-initialized (memset) or calloc used instead or malloc, otherwise entry fields may have arbitrary data and cause undefined behavior if it's used later in the program.
I would suggest to allocate the memory in one go before the loop (json_array_size), fill the entries as the array is processed, if error occurs do `goto exit`.  In the exit section: if no error occured set `settings->coap.security` to allocated device list, otherwise free the allocated list. This will allow failing fast with single exit point, which helps prevent memory leaks when working with dynamic memory.